include/nbl/asset/ICPUPipeline.h

```diff
@@ -14,37 +14,13 @@ namespace nbl::asset
 {
 
 // Common Base class for pipelines
-template<typename PipelineNonAssetBase, uint8_t MaxShaderStageCount>
+template<typename PipelineNonAssetBase>
 class ICPUPipeline : public IAsset, public PipelineNonAssetBase
 {
-        using this_t = ICPUPipeline<PipelineNonAssetBase,MaxShaderStageCount>;
+        using this_t = ICPUPipeline<PipelineNonAssetBase>;
+        using shader_info_spec_t = IPipelineBase::SShaderSpecInfo<true>;
 
     public:
-        inline core::smart_refctd_ptr<IAsset> clone(uint32_t _depth = ~0u) const override final
-        {
-            core::smart_refctd_ptr<ICPUPipelineLayout> layout;
-            if (_depth>0u && PipelineNonAssetBase::m_layout)
-				layout = core::smart_refctd_ptr_static_cast<ICPUPipelineLayout>(PipelineNonAssetBase::m_layout->clone(_depth-1u));
-
-            auto cp = clone_impl(std::move(layout));
-            for (auto i=0; i<MaxShaderStageCount; i++)
-            {
-                const auto shader = m_stages[i].shader;
-                if (shader)
-                {
-                    auto stageInfo = m_stages[i].info;
-                    core::smart_refctd_ptr<IShader> newShader;
-                    if (_depth>0u)
-                    {
-                        newShader = core::smart_refctd_ptr_static_cast<IShader>(shader->clone(_depth-1u));
-                        stageInfo.shader = newShader.get();
-                    }
```

Comment:
make a method which handles cloning a single SpecInfo in a pipeline given this clone depth, make it final and protected in this class

Reply:
my bad, make it a method of the SpecInfo